### PR TITLE
(EAI-201): Step-back prompting for MongoDB AI Chatbot

### PR DESCRIPTION
**Pull request by mongodben**
Jira: https://jira.mongodb.org/browse/EAI-201

## Changes

- Generate search query with step-back prompting
- Separate LLM calls to generate search query and extract metadata for better results
- Bundle all context for LLM in the user message

## Notes

- [ ] Will need to update the production Azure OpenAI ChatGPT API version to use the `0610` model w/ function calling. will need to get our deployment updated to have more tokens/minute using that version as well.

### CODE REVIEW

File: packages/chatbot-server-mongodb-public/src/test/testHelpers.ts [hunk 1]
```diff
@@ -45,4 +45,32 @@ export async function makeTestApp(defaultConfigOverrides?: Partial<AppConfig>) {
...
+  if (magnitudeA && magnitudeB) {
+    // https://towardsdatascience.com/how-to-measure-distances-in-machine-learning-13a396aa34ce
+    return dotProduct(a, b) / (magnitudeA * magnitudeB);
+  } else {
+    return 0;
+  }
+}
+
+function dotProduct(a: number[], b: number[]) {
+  let sum = 0;
```
Comment (cbush) [inline packages/chatbot-server-mongodb-public/src/test/testHelpers.ts:62]: Nit: in a reversal of one of your reviews of mine, I think you could stand to get comfortable with `Array.reduce()` and rewrite this as:
```js
return a.reduce((acc, cur, i) => acc + cur * b[i], 0);
```

**Review by cbush — APPROVED** [review]
Amazing...

File: packages/chatbot-server-mongodb-public/src/test/testHelpers.ts (hunk 1 above)
Comment (mongodben) [reply packages/chatbot-server-mongodb-public/src/test/testHelpers.ts:62]: def (this implementation was a copy-pasta)

File: packages/chatbot-server-mongodb-public/src/test/testHelpers.ts [hunk 2]
```diff
@@ -45,4 +45,32 @@ export async function makeTestApp(defaultConfigOverrides?: Partial<AppConfig>) {
...
+  for (let i = 0; i < a.length; i++) {
+    sum += a[i] * b[i];
+  }
+  return sum;
```
Comment (mongodben) [inline packages/chatbot-server-mongodb-public/src/test/testHelpers.ts:66]:
```suggestion
  return a.reduce((acc, cur, i) => acc + cur * b[i], 0);
```

**Comment by mongodben** [comment]
> Amazing...

hopefully 🙏

**Pull request merged**